feat: add purl as a CLI options (#401)
oracle/macaron: docs/source/pages/using.rst

- Obtain a GitLab access token having at least the ``read_repository`` permission and store it into the ``MCN_SELF_HOSTED_GITLAB_TOKEN`` environment variable. For more detailed instructions, see `GitLab documentation <https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html#create-a-personal-access-token>`_.

''''''''''''''''''''''''''''''''''''''''''''''''''''
Providing a PURL string instead of a repository path
''''''''''''''''''''''''''''''''''''''''''''''''''''

> Instead of providing the repository path for the analysis software component, a `PURL <https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst>`_. string
> for the target git repository could be used.

Reviewer: ```suggestion
Instead of providing the repository path to analyze a software component, you can use a `PURL <https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst>`_. string for the target git repository.
```

Author: Fixed in https://github.com/oracle/macaron/pull/401/commits/229e0394f00c7f257f46987c5549b0d72143aa2a

> To begin with, we follow the same steps for the configuration if needed (especially for self-hosted GitLab instances). The PURL string for a git repository could be created using the following format:

Reviewer: ```suggestion
To simplify the examples, we use the same configurations as above if needed (e.g., for the self-hosted GitLab instances). The PURL string for a git repository should have the following format:
```

Author: Fixed in https://github.com/oracle/macaron/pull/401/commits/229e0394f00c7f257f46987c5549b0d72143aa2a

.. code-block::

  pkg:<git_service_domain>/<organization>/<name>

> The list bellow shows examples for the corresponding PURL string for different git repositories:

Reviewer: ```suggestion
The list bellow shows examples for the corresponding PURL strings for different git repositories:
```

Author: Fixed in https://github.com/oracle/macaron/pull/401/commits/229e0394f00c7f257f46987c5549b0d72143aa2a

.. list-table:: Example of PURL strings for git repositories.
   :widths: 50 50
   :header-rows: 1

   * - Repository path
     - PURL string
   * - ``https://github.com/micronaut-projects/micronaut-core``
     - Both ``pkg:github/micronaut-projects/micronaut-core`` and ``pkg:github.com/micronaut-projects/micronaut-core`` are applicable as ``github`` is a pre-defined type as mentioned `here <https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst>`_.
   * - ``https://bitbucket.org/snakeyaml/snakeyaml``
     - Both ``pkg:github/micronaut-projects/micronaut-core`` and ``pkg:github.com/micronaut-projects/micronaut-core`` are applicable as ``bitbucket`` is a pre-defined type as mentioned `here <https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst>`_.
   * - ``https://internal.gitlab.com/foo/bar``
     - ``pkg:internal.gitlab.com/foo/bar``
   * - ``https://gitlab.com/gitlab-org/gitlab``
     - ``pkg:gitlab.com/gitlab-org/gitlab``

> After the PURL string has been obtained, the analysis can be run with:

Reviewer: ```suggestion
Run the analysis using the PURL string as follows:
```

> After the PURL string has been obtained, the analysis could be run with:

Reviewer 2: ```suggestion
After the PURL string has been obtained, the analysis can be run with:
```

Author: Fixed in https://github.com/oracle/macaron/pull/401/commits/7a84f6f91fe54da57436ef5ac9e065a913932769

.. code-block:: shell

  ./run_macaron.sh analyze -purl <purl_string>

> You can also provide the PURL string together with the repository path. In this case, the PURL string will be used as the unique identifier to the analysis target:

Reviewer 2: ```suggestion
You can also provide the PURL string together with the repository path. In this case, the PURL string will be used as the unique identifier for the analysis target:
```

Author: Fixed in https://github.com/oracle/macaron/pull/401/commits/7a84f6f91fe54da57436ef5ac9e065a913932769

.. code-block:: shell

  ./run_macaron.sh analyze -purl <purl_string> -rp <repo_path> -b <branch> -d <digest>

> .. note:: When provide the PURL and the repository path, both branch name and commit digest must be provided as well.

Reviewer 2: ```suggestion
.. note:: When providing the PURL and the repository path, both the branch name and commit digest must be provided as well.
```

Author: Fixed in https://github.com/oracle/macaron/pull/401/commits/7a84f6f91fe54da57436ef5ac9e065a913932769

Macaron's policy engine accepts policies specified in `Datalog <https://en.wikipedia.org/wiki/Datalog>`_. An example policy
can verify if a project and all its dependencies pass certain checks. We use `Soufflé <https://souffle-lang.github.io/index.html>`_
as the Datalog engine in Macaron. Once you run the checks on a target project as described :ref:`here <analyze-action>`,

> the check results will be stored in ``macaron.db`` in the output directory. Because the check results of different software component can be stored into the database at ``macaron.db``, we must specify the target software component in the Datalog policy file to be enforced by the policy engine. There are two ways that we could

Reviewer 2: ```suggestion
the check results will be stored in ``macaron.db`` in the output directory. Because the check results of different software components can be stored in the database at ``macaron.db``, we must specify the target software component in the Datalog policy file to be enforced by the policy engine. There are two ways that we can
```

Author: Fixed in https://github.com/oracle/macaron/pull/401/commits/7a84f6f91fe54da57436ef5ac9e065a913932769

> the check results will be stored in ``macaron.db`` in the output directory. Because the check results of different software components can be stored in the database at ``macaron.db``, we must specify the target software component in the Datalog policy file to be enforced by the policy engine. There are two ways that we can
> specify the target software component in the Datalog policy file:

Reviewer: ```suggestion
the check results will be stored in ``macaron.db`` in the output directory. We pass the check results to the policy engine by providing the path to ``macaron.db`` together with a Datalog policy file to be validated by the policy engine. In the Datalog policy file, we must specify the identifier for the target software component that we are interested in to validate the policy against. These are two ways to specify the target software component in the Datalog policy file:
```

Author: Fixed in https://github.com/oracle/macaron/pull/401/commits/229e0394f00c7f257f46987c5549b0d72143aa2a

#. Using the complete name of the target component (e.g. ``github.com/oracle-quickstart/oci-micronaut``)
#. Using the PURL string of the target component (e.g. ``pkg:github.com/oracle-quickstart/oci-micronaut@<commit_sha>``).

> However, the two methods would mostly share the same policy definition in the Datalog files.

Reviewer: ```suggestion
```

Author: Fixed in https://github.com/oracle/macaron/pull/401/commits/229e0394f00c7f257f46987c5549b0d72143aa2a

.. code-block:: shell

  ./run_macaron.sh verify-policy -o outputs -d outputs/macaron.db --file <policy_file>

In this example, the Datalog policy files for both ways (as mentioned previously) are provided in `oci-micronaut-repo.dl <../_static/examples/oracle-quickstart/oci-micronaut/policies/oci-micronaut-repo.dl>`__ and `oci-micronaut-purl.dl <../_static/examples/oracle-quickstart/oci-micronaut/policies/oci-micronaut-purl.dl>`__.

The differences between the two policy files can be observed below:

.. tabs::

  .. code-tab:: prolog Using repository complete name

    apply_policy_to("oci_micronaut_dependencies", repo_id) :- is_repo(repo_id, "github.com/oracle-quickstart/oci-micronaut", _).

  .. code-tab:: prolog Using PURL string

    apply_policy_to("oci_micronaut_dependencies", component_id) :- is_component(component_id, "<target_software_component_purl>").

> The PURL string for the target software component could be obtained when the :ref:`analyze command <analyze-action>` finishes. For example:

Reviewer 2: ```suggestion
The PURL string for the target software component can be obtained when the :ref:`analyze command <analyze-action>` has finished. For example:
```

Author: Fixed in https://github.com/oracle/macaron/pull/401/commits/7a84f6f91fe54da57436ef5ac9e065a913932769

> The PURL string for the target software component can be obtained when the :ref:`analyze command <analyze-action>` has finished. For example:

Reviewer: ```suggestion
The PURL string for the target software component is printed to the console by the :ref:`analyze command <analyze-action>`. For example:
```

Author: Fixed in https://github.com/oracle/macaron/pull/401/commits/229e0394f00c7f257f46987c5549b0d72143aa2a
